File: src/StatStudio.Engine/Protocol.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace StatStudio.Engine;

/// <summary>One column of the worksheet as raw strings (missing = null or "").</summary>
public sealed class ColumnDto
{
    public string Name { get; set; } = "";
    public List<string?> Cells { get; set; } = new();
    public string? Type { get; set; }
}

/// <summary>The worksheet as a plain string matrix (the SwiftUI grid mirror).</summary>
public sealed class WorksheetDto
{
    public string? Name { get; set; }
    public List<ColumnDto> Columns { get; set; } = new();
}

/// <summary>A request line: { id, op, worksheet?, params? }.</summary>
public sealed class EngineRequest
{
    public int Id { get; set; }
    public string Op { get; set; } = "";
    public WorksheetDto? Worksheet { get; set; }
    public JsonElement? Params { get; set; }
}

/// <summary>Name + description of a built-in sample dataset (for the File ▸ Sample Data menu).</summary>
public sealed class SampleInfo
{
    public string Name { get; set; } = "";
    public string Description { get; set; } = "";
}

/// <summary>A titled graph image (base64 PNG). An analysis may emit several.</summary>
public sealed class GraphImage
{
    public string Title { get; set; } = "";
    public string Png { get; set; } = "";
}

/// <summary>A response line. Any of sessionText / worksheet / graphs / samples may be set.</summary>
public sealed class EngineResponse
{
    public int Id { get; set; }
    public bool Ok { get; set; }
    public string? StatusTitle { get; set; }
    public string? SessionText { get; set; }
    public WorksheetDto? Worksheet { get; set; }
    public List<GraphImage>? Graphs { get; set; }
    public List<SampleInfo>? Samples { get; set; }
    public string? Error { get; set; }
}

/// <summary>Shared JSON settings: camelCase out, case-insensitive in, skip nulls.</summary>
public static class EngineJson
{
    public static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };
}

/// <summary>Session-text formatting mirroring MainWindow.Output / OutputRaw.</summary>
internal static class Out
{
    /// <summary>Titled block: blank line, title, underline, body. (cf. MainWindow.Output)</summary>
    public static string Block(string title, string body)
    {
        var underline = new string('─', Math.Max(title.Length, 8));
        return $"\n{title}\n{underline}\n{body.TrimEnd()}\n";
    }

    /// <summary>Pre-formatted block whose first line is its own title. (cf. OutputRaw)</summary>
    public static string Raw(string body) => $"\n{body.TrimEnd()}\n";
}

File: mac/StatStudio/Sources/StatStudio/Engine/Protocol.swift
import Foundation

// Wire types mirroring src/StatStudio.Engine/Protocol.cs (camelCase JSON).

struct ColumnDTO: Codable, Sendable {
    var name: String
    var cells: [String?]
    var type: String? = nil
}

struct WorksheetDTO: Codable, Sendable {
    var name: String?
    var columns: [ColumnDTO]
}

struct EngineRequest: Encodable, Sendable {
    var id: Int
    var op: String
    var worksheet: WorksheetDTO?
    var params: [String: JSONValue]?
}

struct SampleInfo: Codable, Sendable, Identifiable {
    var name: String
    var description: String
    var id: String { name }
}

struct GraphImageDTO: Codable, Sendable {
    var title: String
    var png: String
}

struct EngineResponse: Decodable, Sendable {
    var id: Int
    var ok: Bool
    var statusTitle: String?
    var sessionText: String?
    var worksheet: WorksheetDTO?
    var graphs: [GraphImageDTO]?
    var samples: [SampleInfo]?
    var error: String?
}

/// A minimal Codable JSON value, used for the flexible per-op `params` payload.
enum JSONValue: Codable, Sendable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case array([JSONValue])
    case object([String: JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null }
        else if let b = try? c.decode(Bool.self) { self = .bool(b) }
        else if let i = try? c.decode(Int.self) { self = .int(i) }
        else if let d = try? c.decode(Double.self) { self = .double(d) }
        else if let s = try? c.decode(String.self) { self = .string(s) }
        else if let a = try? c.decode([JSONValue].self) { self = .array(a) }
        else if let o = try? c.decode([String: JSONValue].self) { self = .object(o) }
        else { throw DecodingError.dataCorruptedError(in: c, debugDescription: "unsupported JSON value") }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .int(let i): try c.encode(i)
        case .double(let d): try c.encode(d)
        case .bool(let b): try c.encode(b)
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        case .null: try c.encodeNil()
        }
    }
}

extension JSONValue {
    /// Convenience for building a `["columns": .strings([...])]` param.
    static func strings(_ xs: [String]) -> JSONValue { .array(xs.map { .string($0) }) }
}
